servlet/src/main/java/io/grpc/servlet/ServerStream.java

```diff
@@ -0,0 +1,398 @@
+  final Sink sink;
+  final AsyncContext asyncCtx;
+  final AtomicReference<WriteState> writeState;
+  final WritableBufferChain writeChain;
+  final ScheduledExecutorService scheduler;
+
+  ServerStream(
+      WritableBufferAllocator bufferAllocator, AsyncContext asyncCtx,
+      AtomicReference<WriteState> writeState, WritableBufferChain writeChain,
+      ScheduledExecutorService scheduler) {
+    super(bufferAllocator, StatsTraceContext.NOOP);
+    this.asyncCtx = asyncCtx;
+    this.writeState = writeState;
+    this.writeChain = writeChain;
+    this.scheduler = scheduler;
+    this.sink = new Sink();
+  }
+
+  @Override
+  protected TransportState transportState() {
+    return transportState;
+  }
+
+  @Override
+  protected Sink abstractServerStreamSink() {
+    return sink;
+  }
+
+  static final class TransportState extends io.grpc.internal.AbstractServerStream.TransportState {
+
+    final SerializingExecutor transportThreadExecutor =
+        new SerializingExecutor(MoreExecutors.directExecutor());
+
+    TransportState(
+        int maxMessageSize, StatsTraceContext statsTraceCtx, TransportTracer transportTracer) {
+      super(maxMessageSize, statsTraceCtx, transportTracer);
+    }
+
+    @Override
+    public void runOnTransportThread(Runnable r) {
+      transportThreadExecutor.execute(r);
+    }
+
+    @Override
+    public void bytesRead(int numBytes) {
+      // no-op
+      // TODO: flow control
+    }
+
+    @Override
+    public void deframeFailed(Throwable cause) {
+      // TODO
+      cause.printStackTrace();
+    }
+
+    @Override
+    public void inboundDataReceived(ReadableBuffer frame, boolean endOfStream) {
+      runOnTransportThread(() -> super.inboundDataReceived(frame, endOfStream));
+    }
+  }
+
+  static final class ByteArrayWritableBuffer implements WritableBuffer {
+
+    static final ByteArrayWritableBuffer FLUSH = new ByteArrayWritableBuffer();
+
+    private final int capacity;
+    final byte[] bytes;
+    private int index;
+
+    private ByteArrayWritableBuffer() {
+      capacity = 0;
+      bytes = new byte[0];
+    }
+
+    ByteArrayWritableBuffer(int capacityHint) {
+      capacity = min(1024 * 1024,  max(4096, capacityHint));
+      bytes = new byte[capacity];
+    }
+
+    @Override
+    public void write(byte[] src, int srcIndex, int length) {
+      System.arraycopy(src, srcIndex, bytes, index, length);
+      index += length;
+    }
+
+    @Override
+    public void write(byte b) {
+      bytes[index++] = b;
+    }
+
+    @Override
+    public int writableBytes() {
+      return capacity - index;
+    }
+
+    @Override
+    public int readableBytes() {
+      return index;
+    }
+
+    @Override
+    public void release() {}
+  }
+
+  /**
+   * A queue of WritableBuffers. Not safe for multiple concurrent polls, or multiple concurrent
+   * enqueues, but safe for concurrently calling one poll() and one enqueue().
+   */
+  static final class WritableBufferChain {
```

Comment:
Why not use LinkedBlockingQueue?

Reply:
Because I just need a simple linked list, only requirement is next pointer be up-to-date, no compare-and-swap, no loop.

Reply:
Those are all things you don't need. What do you need? You wanted it to be faster?

You need a thread-safe queue. `LinkedBlockingQueue` is the canonical one, is fewer lines of code to use, and less likely to have a bug (other than maybe it having a more complex API).

Reply:
Why do you suggest `LinkedBlockingQueue`? Even `ConcurrentLinkedQueue` would be enough. I don't need the thread-safety it provides. The thread-safety is handled at the its usage outside, the thread-safety by the queue itself does not help and is still not enough.

Reply:
`ConcurrentLinkedQueue` is fine. I'm mainly arguing to not write your own unless you get some benefit from it.

Reply:
If you don't need thread-safety, then ArrayDeque or LinkedList would be appropriate. Both implement `java.util.Queue`.

Reply:
The concurrency is handled by the `writeState` state machine. Unfortunately it's only capable of handling the concurrency for singly-linked queue. Deque or doubly-linked queue would messed up the state. `ArrayDeque` would work if there were no `doubleCapacity()`. So from `java.util.*`  only `ConcurrentLinkedQueue` can be used.